<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Run sciMARGI-tool pipeline in Docker &mdash; sciMARGI-tool 0.0.1 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Data processing step-by-step" href="step.html" />
    <link rel="prev" title="sciMARGI Introduction" href="scimargiintro.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> sciMARGI-tool
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="scimargiintro.html">sciMARGI Introduction</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Run sciMARGI-tool pipeline in Docker</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#install-docker-on-your-system">Install Docker on your system</a></li>
<li class="toctree-l2"><a class="reference internal" href="#install-scimargi-tool-image">Install sciMARGI-tool image</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run-scimargi-docker">Run sciMARGI-docker</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#data-preparation">Data preparation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#initiate-the-scimargi-tool-image">Initiate the sciMARGI-tool image</a></li>
<li class="toctree-l3"><a class="reference internal" href="#running-time-of-the-pipeline">Running time of the pipeline</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#output-of-scimargi-docker-pipeline">Output of sciMARGI-docker pipeline</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="step.html">Data processing step-by-step</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">sciMARGI-tool</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Run sciMARGI-tool pipeline in Docker</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="run-scimargi-tool-pipeline-in-docker">
<h1>Run sciMARGI-tool pipeline in Docker<a class="headerlink" href="#run-scimargi-tool-pipeline-in-docker" title="Permalink to this heading">¶</a></h1>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this heading">¶</a></h2>
<a class="reference internal image-reference" href="../_images/sciMARGI_pipeline.png"><img alt="sciMARGI_docker_pipeline" src="../_images/sciMARGI_pipeline.png" style="width: 850px;" /></a>
<p>sciMARGI-docker is a tool designed to provide a one-stop solution for processing sciMARGI data. It is a docker image that wraps a Snakemake pipeline. sciMARGI-docker combines the power of Docker and Snakemake to provide a one-stop solution for processing sciMARGI data. It allows you to easily deploy and run the sciMARGI pipeline in a reproducible and efficient way, making it easier to manage and automate the data processing process.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>Why we choose docker?</strong></p>
<p>Docker is a tool for packaging and deploying applications in lightweight containers. It allows you to package an application and its dependencies in a single container, making it easy to deploy and run on any platform. The philosophy behind Docker is to make it easier to create, deploy, and run applications by packaging them in containers. You can find more information about Docker and how to use it in the official Docker documentation at <a class="reference external" href="https://docs.docker.com/">https://docs.docker.com/</a>.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>Why we choose Snakemake?</strong></p>
<p>Snakemake is a workflow management system that allows you to define and run complex pipelines in a reproducible and efficient way. It is designed to be scalable, flexible, and easy to use, with a simple yet powerful syntax for defining workflows. The philosophy behind Snakemake is to provide a simple and intuitive way to define and run complex pipelines, making it easier to manage and automate complex workflow processes. You can find more information about Snakemake and how to use it in the official Snakemake documentation at <a class="reference external" href="https://snakemake.readthedocs.io/">https://snakemake.readthedocs.io/</a>.</p>
</div>
</section>
<section id="install-docker-on-your-system">
<h2>Install Docker on your system<a class="headerlink" href="#install-docker-on-your-system" title="Permalink to this heading">¶</a></h2>
<p>If you don’t have docker installed, please follow the instructions from docker docs: <a class="reference external" href="https://docs.docker.com/engine/install/">https://docs.docker.com/engine/install/</a></p>
</section>
<section id="install-scimargi-tool-image">
<h2>Install sciMARGI-tool image<a class="headerlink" href="#install-scimargi-tool-image" title="Permalink to this heading">¶</a></h2>
<p>sciMARGI-tool has now been uploaded to docker cloud, which means you can install the container  by <code class="docutils literal notranslate"><span class="pre">docker</span> <span class="pre">pull</span></code> command:</p>
<div class="highlight highlight-bash notranslate"><div class="highlight"><pre><span></span>docker pull zhonglab/scimargi
</pre></div>
</div>
</section>
<section id="run-scimargi-docker">
<h2>Run sciMARGI-docker<a class="headerlink" href="#run-scimargi-docker" title="Permalink to this heading">¶</a></h2>
<section id="data-preparation">
<h3>Data preparation<a class="headerlink" href="#data-preparation" title="Permalink to this heading">¶</a></h3>
<p>To run sciMARGI-tool locally, we first need to prepare the following input files in a structured directory. The docker image (created from sciMARGI-tool container) will run in an independent virtual enviornment. The only way for the image to communicate with your local file is through the <code class="docutils literal notranslate"><span class="pre">mount</span></code> command. Softwares running inside the docker image will recognize input files with the path inside the docker enviornment but not the file path on your local machine. So it is cruial to keep your input files in a structured directory and pass down parameters using relative path.</p>
<p><strong>Required input files:</strong></p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>File format</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>File name</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>.fastq.gz</p></td>
<td><p>Input raw fastq files in gzipped format. Should have Read1 and Read2 files for 8 libs (if the initial library is spliited into 8 aliqotes and using unique index before sequencing).</p></td>
<td><p>Not fixed, as long as it is in gzipped format.</p></td>
</tr>
<tr class="row-odd"><td><p>.json</p></td>
<td><p>A JSON file to specify the name of each lib and the file path of corresponded R1, R2 files. Notice the file path should only use relative path to the JSON file.</p></td>
<td><p>Required and fixed: samples.json</p></td>
</tr>
<tr class="row-even"><td><p>.gtf</p></td>
<td><p>gtf file of the reference genome of interest.</p></td>
<td><p>Not fixed</p></td>
</tr>
<tr class="row-odd"><td><p>.fa</p></td>
<td><p>fasta file of the reference genome of interest.</p></td>
<td><p>Not fixed</p></td>
</tr>
</tbody>
</table>
<p>On your local machine prepare a input folder structured as follows:</p>
<div class="highlight highlight-bash notranslate"><div class="highlight"><pre><span></span>&gt; tree local_path/sciMARGI_input_folder/

├── data
│   ├── lib1_S1_R1_001.fastq.gz
│   ├── lib1_S1_R2_001.fastq.gz
│   ├── lib2_S2_R1_001.fastq.gz
│   ├── lib2_S2_R2_001.fastq.gz
│   ├── lib3_S3_R1_001.fastq.gz
│   ├── lib3_S3_R2_001.fastq.gz
│   ├── lib4_S4_R1_001.fastq.gz
│   ├── lib4_S4_R2_001.fastq.gz
│   ├── lib5_S5_R1_001.fastq.gz
│   ├── lib5_S5_R2_001.fastq.gz
│   ├── lib6_S6_R1_001.fastq.gz
│   ├── lib6_S6_R2_001.fastq.gz
│   ├── lib7_S7_R1_001.fastq.gz
│   ├── lib7_S7_R2_001.fastq.gz
│   ├── lib8_S8_R1_001.fastq.gz
│   ├── lib8_S8_R2_001.fastq.gz
│   └── samples.json
├── output
└── ref
    ├── gencode.v36.annotation.gtf
    └── hg38_genome.fa
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Notice that the output folder is necessary (even though it is empty right now), and it is the folder where your final output files will be located.</p>
</div>
<p>The samples.json file in our example will look like</p>
<div class="highlight highlight-bash notranslate"><div class="highlight"><pre><span></span>// samples.json
<span class="o">{</span>
  <span class="s2">&quot;lib1&quot;</span>: <span class="o">{</span>
    <span class="s2">&quot;Sample&quot;</span>: <span class="s2">&quot;lib1&quot;</span>,
    <span class="s2">&quot;R1&quot;</span>: <span class="s2">&quot;lib1_S1_R1_001.fastq.gz&quot;</span>, <span class="c1"># path to R1 file.</span>
    <span class="s2">&quot;R2&quot;</span>: <span class="s2">&quot;lib1_S1_R2_001.fastq.gz&quot;</span>
  <span class="o">}</span>,
  <span class="s2">&quot;lib2&quot;</span>: <span class="o">{</span>
    <span class="s2">&quot;Sample&quot;</span>: <span class="s2">&quot;lib2&quot;</span>,
    <span class="s2">&quot;R1&quot;</span>: <span class="s2">&quot;lib2_S2_R1_001.fastq.gz&quot;</span>,
    <span class="s2">&quot;R2&quot;</span>: <span class="s2">&quot;lib2_S2_R2_001.fastq.gz&quot;</span>
  <span class="o">}</span>,
  <span class="s2">&quot;lib3&quot;</span>: <span class="o">{</span>
    <span class="s2">&quot;Sample&quot;</span>: <span class="s2">&quot;lib3&quot;</span>,
    <span class="s2">&quot;R1&quot;</span>: <span class="s2">&quot;lib3_S3_R1_001.fastq.gz&quot;</span>,
    <span class="s2">&quot;R2&quot;</span>: <span class="s2">&quot;lib3_S3_R2_001.fastq.gz&quot;</span>
  <span class="o">}</span>,
  <span class="s2">&quot;lib4&quot;</span>: <span class="o">{</span>
    <span class="s2">&quot;Sample&quot;</span>: <span class="s2">&quot;lib4&quot;</span>,
    <span class="s2">&quot;R1&quot;</span>: <span class="s2">&quot;lib4_S4_R1_001.fastq.gz&quot;</span>,
    <span class="s2">&quot;R2&quot;</span>: <span class="s2">&quot;lib4_S4_R2_001.fastq.gz&quot;</span>
  <span class="o">}</span>,
  <span class="s2">&quot;lib5&quot;</span>: <span class="o">{</span>
    <span class="s2">&quot;Sample&quot;</span>: <span class="s2">&quot;lib5&quot;</span>,
    <span class="s2">&quot;R1&quot;</span>: <span class="s2">&quot;lib5_S5_R1_001.fastq.gz&quot;</span>,
    <span class="s2">&quot;R2&quot;</span>: <span class="s2">&quot;lib5_S5_R2_001.fastq.gz&quot;</span>
  <span class="o">}</span>,
  <span class="s2">&quot;lib6&quot;</span>: <span class="o">{</span>
    <span class="s2">&quot;Sample&quot;</span>: <span class="s2">&quot;lib6&quot;</span>,
    <span class="s2">&quot;R1&quot;</span>: <span class="s2">&quot;lib6_S6_R1_001.fastq.gz&quot;</span>,
    <span class="s2">&quot;R2&quot;</span>: <span class="s2">&quot;lib6_S6_R2_001.fastq.gz&quot;</span>
  <span class="o">}</span>,
  <span class="s2">&quot;lib7&quot;</span>: <span class="o">{</span>
    <span class="s2">&quot;Sample&quot;</span>: <span class="s2">&quot;lib7&quot;</span>,
    <span class="s2">&quot;R1&quot;</span>: <span class="s2">&quot;lib7_S7_R1_001.fastq.gz&quot;</span>,
    <span class="s2">&quot;R2&quot;</span>: <span class="s2">&quot;lib7_S7_R2_001.fastq.gz&quot;</span>
  <span class="o">}</span>,
  <span class="s2">&quot;lib8&quot;</span>: <span class="o">{</span>
    <span class="s2">&quot;Sample&quot;</span>: <span class="s2">&quot;lib8&quot;</span>,
    <span class="s2">&quot;R1&quot;</span>: <span class="s2">&quot;lib8_S8_R1_001.fastq.gz&quot;</span>,
    <span class="s2">&quot;R2&quot;</span>: <span class="s2">&quot;lib8_S8_R2_001.fastq.gz&quot;</span>
  <span class="o">}</span>
<span class="o">}</span>
</pre></div>
</div>
<p>These are all required input files for sciMARGI-tool to run!</p>
</section>
<section id="initiate-the-scimargi-tool-image">
<h3>Initiate the sciMARGI-tool image<a class="headerlink" href="#initiate-the-scimargi-tool-image" title="Permalink to this heading">¶</a></h3>
<p>After all input files are ready, we are ready to initiate the pipelline with <code class="docutils literal notranslate"><span class="pre">docker</span> <span class="pre">run</span></code> command.</p>
<div class="highlight highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># download sciMARGI-tool docker image:</span>
docker pull zhonglab/scimargi:latest

<span class="c1"># run the pipeline</span>
docker run --rm <span class="se">\</span>
        -v /local_path/sciMARGI_input_folder/:/input <span class="se">\</span>
        -v /local_path/sciMARGI_input_folder/output/:/snakemake/outputs <span class="se">\</span>
        scimargi:latest <span class="se">\</span>
        --sample-json /input/data/samples.json <span class="se">\</span>
        --num-jobs <span class="m">20</span> <span class="se">\</span>
        --name mysample <span class="se">\</span>
        --fa /input/ref/hg38_genome.fa <span class="se">\</span>
        --gtf /input/ref/gencode.v36.annotation.gtf <span class="se">\</span>
        --uid <span class="sb">`</span>id -u<span class="sb">`</span> <span class="se">\</span>
        --gid <span class="sb">`</span>id -g<span class="sb">`</span>
</pre></div>
</div>
<p><em>Parameters explaination:</em></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--rm</span></code> : By default, when a container exits, it is not removed and can still be inspected and restarted using the <code class="docutils literal notranslate"><span class="pre">docker</span> <span class="pre">start</span></code> command. However, if the <code class="docutils literal notranslate"><span class="pre">--rm</span></code> flag is specified, the container will be automatically removed once it exits. This can be useful for cleaning up after a one-time task, as it ensures that the system does not accumulate stopped containers that are no longer needed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-v</span></code>:  The <cite>-v</cite> flag is used with the <cite>docker run</cite> command to bind-mount a volume from the host system into a container. This allows you to share data between the host system and the container, or to persist data that is generated within the container beyond the lifetime of the container. the host path and the container path separated by a colon (<cite>:</cite>). In our command, we mount twice, the first one is <code class="docutils literal notranslate"><span class="pre">`</span> <span class="pre">-v</span> <span class="pre">/local_path/sciMARGI_input_folder/:/input</span> <span class="pre">`</span></code> which mount the local input folder to folder <code class="docutils literal notranslate"><span class="pre">`/input`</span></code> in docker image. As you could see, later when we specify other parameters such as <code class="docutils literal notranslate"><span class="pre">`-fa,</span> <span class="pre">-gtf`</span></code>, we will use the new path that is in docker image as <code class="docutils literal notranslate"><span class="pre">`</span> <span class="pre">/input/ref/*.fa</span> <span class="pre">`</span></code>.  The second one is <code class="docutils literal notranslate"><span class="pre">`</span> <span class="pre">-v</span> <span class="pre">/local_path/sciMARGI_input_folder/out/:/snakemake/outputs</span> <span class="pre">`</span></code> which is used to map docker produced output back to our local folder.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">scimargi:latest</span></code>: Before adding other customized parameters, we need to specify the image name, which is usually <cite>name:tag</cite></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--sample-json</span></code>: Location of the samples.json file inside the docker, not local file path. No need to change the default if you structured your input folder as we suggested.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--num-jobs</span></code>: Threads for running the job. Snakemake will automatically distribute cpus to all jobs in a very efficient manner.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--fa</span></code>: Location of the reference genome fasta file inside the docker, not local file path. No need to change the default if you structured your input folder as we suggested.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--gtf</span></code>: Similar to <code class="docutils literal notranslate"><span class="pre">--fa</span></code>, but for location of the reference genome gtf file.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--uid</span> <span class="pre">'id</span> <span class="pre">-u'</span></code>:  No need to change this command. Here <code class="docutils literal notranslate"><span class="pre">id</span> <span class="pre">-u</span></code> will obtain your user id on your local machine. This parameter is extremely crucial and it allows you (local user) to have permissions and ownership to all output files that produced by docker and Snakemake. Docker by default will create all files under <code class="docutils literal notranslate"><span class="pre">USER</span> <span class="pre">root</span></code> which is extremely inconvenient. However, the change of ownership from root to local user will only be done after all output has finished. If your program is killed or stopped before finish the pipeline, the output file can still be owner by root. Notice it’s backtick wraps around <code class="docutils literal notranslate"><span class="pre">id</span> <span class="pre">-u</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--uid</span> <span class="pre">'id</span> <span class="pre">-g'</span></code>: No need to change this command. Here <code class="docutils literal notranslate"><span class="pre">id</span> <span class="pre">-g</span></code> will obtain your group id on your local machine.</p></li>
</ul>
</section>
<section id="running-time-of-the-pipeline">
<h3>Running time of the pipeline<a class="headerlink" href="#running-time-of-the-pipeline" title="Permalink to this heading">¶</a></h3>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Add benchmark with test data.</p>
</div>
</section>
</section>
<section id="output-of-scimargi-docker-pipeline">
<h2>Output of sciMARGI-docker pipeline<a class="headerlink" href="#output-of-scimargi-docker-pipeline" title="Permalink to this heading">¶</a></h2>
<p>The output folder includes all intermediate files for users to check and potential making use of. Briefly we have 6 major folder corresponds to each step. Detailed description of each step can be found in <a class="reference external" href="./step.rst">Data processing step-by-step</a></p>
<p>The full tree structure of the final output folder will look like:</p>
<div class="highlight highlight-bash notranslate"><div class="highlight"><pre><span></span>&gt; tree /local_path/sciMARGI_input_folder/output/

├── 0_index
│   ├── bowtie2
│   │   ├── user.1.bt2
│   │   ├── user.2.bt2
│   │   ├── user.3.bt2
│   │   ├── user.4.bt2
│   │   ├── user_index_done.txt
│   │   ├── user.rev.1.bt2
│   │   └── user.rev.2.bt2
│   └── STAR
│       ├── chrLength.txt
│       ├── chrNameLength.txt
│       ├── chrName.txt
│       ├── chrStart.txt
│       ├── exonGeTrInfo.tab
│       ├── exonInfo.tab
│       ├── geneInfo.tab
│       ├── Genome
│       ├── genomeParameters.txt
│       ├── SA
│       ├── SAindex
│       ├── sjdbInfo.txt
│       ├── sjdbList.fromGTF.out.tab
│       ├── sjdbList.out.tab
│       └── transcriptInfo.tab
├── 1_decode
│   ├── lib1_cutadap_trimed_DNAend_decode.fq
│   ├── lib1_cutadap_trimed_RNAend_decode.fq
│   ├── lib1_DNAend_decode.fq
│   ├── lib1_RNAend_decode.fq
│   ├── lib1_stats.log
│   ├── lib2_cutadap_trimed_DNAend_decode.fq
│   ├── lib2_cutadap_trimed_RNAend_decode.fq
│   ├── lib2_DNAend_decode.fq
│   ├── lib2_RNAend_decode.fq
│   ├── lib2_stats.log
│   ├── lib3_cutadap_trimed_DNAend_decode.fq
│   ├── lib3_cutadap_trimed_RNAend_decode.fq
│   ├── lib3_DNAend_decode.fq
│   ├── lib3_RNAend_decode.fq
│   ├── lib3_stats.log
│   ├── lib4_cutadap_trimed_DNAend_decode.fq
│   ├── lib4_cutadap_trimed_RNAend_decode.fq
│   ├── lib4_DNAend_decode.fq
│   ├── lib4_RNAend_decode.fq
│   ├── lib4_stats.log
│   ├── lib5_cutadap_trimed_DNAend_decode.fq
│   ├── lib5_cutadap_trimed_RNAend_decode.fq
│   ├── lib5_DNAend_decode.fq
│   ├── lib5_RNAend_decode.fq
│   ├── lib5_stats.log
│   ├── lib6_cutadap_trimed_DNAend_decode.fq
│   ├── lib6_cutadap_trimed_RNAend_decode.fq
│   ├── lib6_DNAend_decode.fq
│   ├── lib6_RNAend_decode.fq
│   ├── lib6_stats.log
│   ├── lib7_cutadap_trimed_DNAend_decode.fq
│   ├── lib7_cutadap_trimed_RNAend_decode.fq
│   ├── lib7_DNAend_decode.fq
│   ├── lib7_RNAend_decode.fq
│   ├── lib7_stats.log
│   ├── lib8_cutadap_trimed_DNAend_decode.fq
│   ├── lib8_cutadap_trimed_RNAend_decode.fq
│   ├── lib8_DNAend_decode.fq
│   ├── lib8_RNAend_decode.fq
│   └── lib8_stats.log
├── 2_DNAmapping
│   ├── human_bowtie2_mapq20_lib1.sorted.bam
│   ├── human_bowtie2_mapq20_lib1.stats.txt
│   ├── human_bowtie2_mapq20_lib2.sorted.bam
│   ├── human_bowtie2_mapq20_lib2.stats.txt
│   ├── human_bowtie2_mapq20_lib3.sorted.bam
│   ├── human_bowtie2_mapq20_lib3.stats.txt
│   ├── human_bowtie2_mapq20_lib4.sorted.bam
│   ├── human_bowtie2_mapq20_lib4.stats.txt
│   ├── human_bowtie2_mapq20_lib5.sorted.bam
│   ├── human_bowtie2_mapq20_lib5.stats.txt
│   ├── human_bowtie2_mapq20_lib6.sorted.bam
│   ├── human_bowtie2_mapq20_lib6.stats.txt
│   ├── human_bowtie2_mapq20_lib7.sorted.bam
│   ├── human_bowtie2_mapq20_lib7.stats.txt
│   ├── human_bowtie2_mapq20_lib8.sorted.bam
│   └── human_bowtie2_mapq20_lib8.stats.txt
├── 3_RNAmapping
│   ├── lib1
│   │   ├── human_STAR_mapq255_lib1.bam
│   │   ├── human_STAR_mapq255_lib1.bam.bai
│   │   ├── lib1_human_Aligned.sortedByCoord.out.bam
│   │   ├── lib1_human_Log.final.out
│   │   ├── lib1_human_Log.out
│   │   ├── lib1_human_Log.progress.out
│   │   ├── lib1_human_SJ.out.tab
│   │   └── lib1_human__STARtmp
│   │       └── BAMsort
│   │           ├── <span class="m">0</span>
│   │           ├── <span class="m">1</span>
│   │           ├── <span class="m">2</span>
│   │           ├── <span class="m">3</span>
│   │           └── <span class="m">4</span>
│   ├── lib2
│   │   ├── human_STAR_mapq255_lib2.bam
│   │   ├── human_STAR_mapq255_lib2.bam.bai
│   │   ├── lib2_human_Aligned.sortedByCoord.out.bam
│   │   ├── lib2_human_Log.final.out
│   │   ├── lib2_human_Log.out
│   │   ├── lib2_human_Log.progress.out
│   │   ├── lib2_human_SJ.out.tab
│   │   └── lib2_human__STARtmp
│   │       └── BAMsort
│   │           ├── <span class="m">0</span>
│   │           ├── <span class="m">1</span>
│   │           ├── <span class="m">2</span>
│   │           ├── <span class="m">3</span>
│   │           └── <span class="m">4</span>
│   ├── lib3
│   │   ├── human_STAR_mapq255_lib3.bam
│   │   ├── human_STAR_mapq255_lib3.bam.bai
│   │   ├── lib3_human_Aligned.sortedByCoord.out.bam
│   │   ├── lib3_human_Log.final.out
│   │   ├── lib3_human_Log.out
│   │   ├── lib3_human_Log.progress.out
│   │   ├── lib3_human_SJ.out.tab
│   │   └── lib3_human__STARtmp
│   │       └── BAMsort
│   │           ├── <span class="m">0</span>
│   │           ├── <span class="m">1</span>
│   │           ├── <span class="m">2</span>
│   │           ├── <span class="m">3</span>
│   │           └── <span class="m">4</span>
│   ├── lib4
│   │   ├── human_STAR_mapq255_lib4.bam
│   │   ├── human_STAR_mapq255_lib4.bam.bai
│   │   ├── lib4_human_Aligned.sortedByCoord.out.bam
│   │   ├── lib4_human_Log.final.out
│   │   ├── lib4_human_Log.out
│   │   ├── lib4_human_Log.progress.out
│   │   ├── lib4_human_SJ.out.tab
│   │   └── lib4_human__STARtmp
│   │       └── BAMsort
│   │           ├── <span class="m">0</span>
│   │           ├── <span class="m">1</span>
│   │           ├── <span class="m">2</span>
│   │           ├── <span class="m">3</span>
│   │           └── <span class="m">4</span>
│   ├── lib5
│   │   ├── human_STAR_mapq255_lib5.bam
│   │   ├── human_STAR_mapq255_lib5.bam.bai
│   │   ├── lib5_human_Aligned.sortedByCoord.out.bam
│   │   ├── lib5_human_Log.final.out
│   │   ├── lib5_human_Log.out
│   │   ├── lib5_human_Log.progress.out
│   │   ├── lib5_human_SJ.out.tab
│   │   └── lib5_human__STARtmp
│   │       └── BAMsort
│   │           ├── <span class="m">0</span>
│   │           ├── <span class="m">1</span>
│   │           ├── <span class="m">2</span>
│   │           ├── <span class="m">3</span>
│   │           └── <span class="m">4</span>
│   ├── lib6
│   │   ├── human_STAR_mapq255_lib6.bam
│   │   ├── human_STAR_mapq255_lib6.bam.bai
│   │   ├── lib6_human_Aligned.sortedByCoord.out.bam
│   │   ├── lib6_human_Log.final.out
│   │   ├── lib6_human_Log.out
│   │   ├── lib6_human_Log.progress.out
│   │   ├── lib6_human_SJ.out.tab
│   │   └── lib6_human__STARtmp
│   │       └── BAMsort
│   │           ├── <span class="m">0</span>
│   │           ├── <span class="m">1</span>
│   │           ├── <span class="m">2</span>
│   │           ├── <span class="m">3</span>
│   │           └── <span class="m">4</span>
│   ├── lib7
│   │   ├── human_STAR_mapq255_lib7.bam
│   │   ├── human_STAR_mapq255_lib7.bam.bai
│   │   ├── lib7_human_Aligned.sortedByCoord.out.bam
│   │   ├── lib7_human_Log.final.out
│   │   ├── lib7_human_Log.out
│   │   ├── lib7_human_Log.progress.out
│   │   ├── lib7_human_SJ.out.tab
│   │   └── lib7_human__STARtmp
│   │       └── BAMsort
│   │           ├── <span class="m">0</span>
│   │           ├── <span class="m">1</span>
│   │           ├── <span class="m">2</span>
│   │           ├── <span class="m">3</span>
│   │           └── <span class="m">4</span>
│   └── lib8
│       ├── human_STAR_mapq255_lib8.bam
│       ├── human_STAR_mapq255_lib8.bam.bai
│       ├── lib8_human_Aligned.sortedByCoord.out.bam
│       ├── lib8_human_Log.final.out
│       ├── lib8_human_Log.out
│       ├── lib8_human_Log.progress.out
│       ├── lib8_human_SJ.out.tab
│       └── lib8_human__STARtmp
│           └── BAMsort
│               ├── <span class="m">0</span>
│               ├── <span class="m">1</span>
│               ├── <span class="m">2</span>
│               ├── <span class="m">3</span>
│               └── <span class="m">4</span>
├── 4_dedup
│   ├── lib1
│   │   ├── human_bowtie2_mapq20_lib1.sorted.dedup.bam
│   │   └── human_RNA_STAR_mapq255_lib1.sorted.dedup.bam
│   ├── lib2                                                                                                                                                    <span class="o">[</span><span class="m">0</span>/432<span class="o">]</span>
│   │   ├── human_bowtie2_mapq20_lib2.sorted.dedup.bam
│   │   └── human_RNA_STAR_mapq255_lib2.sorted.dedup.bam
│   ├── lib3
│   │   ├── human_bowtie2_mapq20_lib3.sorted.dedup.bam
│   │   └── human_RNA_STAR_mapq255_lib3.sorted.dedup.bam
│   ├── lib4
│   │   ├── human_bowtie2_mapq20_lib4.sorted.dedup.bam
│   │   └── human_RNA_STAR_mapq255_lib4.sorted.dedup.bam
│   ├── lib5
│   │   ├── human_bowtie2_mapq20_lib5.sorted.dedup.bam
│   │   └── human_RNA_STAR_mapq255_lib5.sorted.dedup.bam
│   ├── lib6
│   │   ├── human_bowtie2_mapq20_lib6.sorted.dedup.bam
│   │   └── human_RNA_STAR_mapq255_lib6.sorted.dedup.bam
│   ├── lib7
│   │   ├── human_bowtie2_mapq20_lib7.sorted.dedup.bam
│   │   └── human_RNA_STAR_mapq255_lib7.sorted.dedup.bam
│   └── lib8
│       ├── human_bowtie2_mapq20_lib8.sorted.dedup.bam
│       └── human_RNA_STAR_mapq255_lib8.sorted.dedup.bam
├── 5_merge
│   ├── merge_DNA_human.sort.bam
│   └── merge_RNA_human.sort.bam
├── 6_stats
│   ├── merge_DNA_human.sort.cell_clusters.csv
│   ├── merge_DNA_human.sort.cell_reads.csv
│   ├── merge_DNA_human.sort.clusters_size.csv
│   ├── merge_human.sort.cluster_rna_dna.csv
│   ├── merge_RNA_human.sort.cell_clusters.csv
│   ├── merge_RNA_human.sort.cell_reads.csv
│   └── merge_RNA_human.sort.clusters_size.csv
└── Library_stats.docx
</pre></div>
</div>
<p></p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="scimargiintro.html" class="btn btn-neutral float-left" title="sciMARGI Introduction" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="step.html" class="btn btn-neutral float-right" title="Data processing step-by-step" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Zhong Lab, UCSD. Author: Xingzhao Wen.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>